network-api/networkapi/wagtailpages/pagemodels/blog_index.py

```diff
@@ -0,0 +1,111 @@
+from django.core.exceptions import ObjectDoesNotExist
+from django.shortcuts import redirect
+
+from wagtail.contrib.routable_page.models import route
+
+from networkapi.wagtailpages.utils import titlecase
+
+from .index import IndexPage
+from .blog import BlogPageCategory
+
+
+class BlogIndexPage(IndexPage):
+    """
+    The blog index is specifically for blog pages,
+    with additional logic to explore categories.
+    """
+
+    subpage_types = [
+        'BlogPage',
+    ]
+
+    template = 'wagtailpages/index_page.html'
+
+    def filter_entries(self, entries, context):
+        entries = super().filter_entries(entries, context)
+
+        if context['filtered'] == 'category':
+            entries = self.filter_entries_for_category(entries, context)
+            context['total_entries'] = len(entries)
+
+        return entries
+
+    def filter_entries_for_category(self, entries, context):
+        category = self.filtered.get('category')
+
+        # make sure we bypass "x results for Y"
+        context['no_filter_ui'] = True
+
+        # and that we don't show the primary tag/category
+        context['hide_classifiers'] = True
+
+        # explicitly set the index page title and intro
+        context['index_title'] = titlecase(f'{category.name} {self.title}')
+        context['index_intro'] = category.intro
+
+        # and then the filtered content
+        context['terms'] = [category.name, ]
+        entries = [
+            entry
+            for
+            entry in entries.specific()
+            if
+            hasattr(entry, 'category')
+            and
+            category in entry.category.all()
+        ]
+
+        return entries
+
+    # helper function to resolve category slugs to actual objects
+    def get_category_object_for_slug(self, category_slug):
+        # We can't use .filter for @property fields,
```

Comment:
removed the line that got the BlogPageCategory model from the database rather than as direct import, because in the old code that would have been a circular dependency, but with these refactors: that's no longer the case and we can import it at the top of this file and it's very nice yes.